# Data Splitting

---

In this phase, we transition from a single dataset to isolated subsets to ensure our model can generalize to unseen data. Since the Challenge Rating (CR) is highly imbalanced—with thousands of low-level "minions" and very few "legendary bosses"—we utilize **Stratified Splitting**. This ensures that the rare high-CR monsters are represented proportionally in both the training and testing phases, preventing the model from becoming biased toward weaker creatures.

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Set plotting style
plt.style.use('fivethirtyeight')
sns.set_theme(style="whitegrid")

# Load preprocessed dataset
preprocessed_df = pd.read_csv('../../data/proposed/preprocessed_bestiary.csv')

# Print dataset information
print("--- Dataset Information ---")
preprocessed_df.info()

# Display summary statistics
print("\n--- Statistical Summary ---")
display(preprocessed_df.describe())

# Display first 5 records
print("\n--- First 5 Records ---")
display(preprocessed_df.head())

--- Dataset Information ---
<class 'pandas.DataFrame'>
RangeIndex: 3658 entries, 0 to 3657
Data columns (total 28 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   CR                 3658 non-null   float64
 1   Size_Rank          3658 non-null   int64  
 2   is_swarm           3658 non-null   int64  
 3   Type_aberration    3658 non-null   int64  
 4   Type_beast         3658 non-null   int64  
 5   Type_construct     3658 non-null   int64  
 6   Type_dragon        3658 non-null   int64  
 7   Type_elemental     3658 non-null   int64  
 8   Type_fiend         3658 non-null   int64  
 9   Type_giant         3658 non-null   int64  
 10  Type_humanoid      3658 non-null   int64  
 11  Type_monstrosity   3658 non-null   int64  
 12  Type_other         3658 non-null   int64  
 13  Type_undead        3658 non-null   int64  
 14  Is_Legendary       3658 non-null   int64  
 15  Is_Spellcaster     3658 non-null   int64  
 16  Log_Avg

,CR,Size_Rank,is_swarm,Type_aberration,Type_beast,Type_construct,Type_dragon,Type_elemental,Type_fiend,Type_giant,...,Strength_Z,Dexterity_Z,Constitution_Z,Intelligence_Z,Wisdom_Z,Charisma_Z,Max_Speed_Z,Trait_Count_MM,Action_Count_MM,Movement_Count_MM
count,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,3658.000000,...,3.658000e+03,3.658000e+03,3.658000e+03,3.658000e+03,3.658000e+03,3.658000e+03,3.658000e+03,3658.000000,3658.000000,3658.000000
mean,5.986844,3.366320,0.010935,0.054128,0.073264,0.073264,0.052214,0.031985,0.065336,0.044013,...,1.398553e-16,1.456826e-18,-3.884870e-17,9.323688e-17,-7.769740e-17,-1.165461e-16,-1.126612e-16,0.132142,0.230215,0.155185
std,6.189269,1.046777,0.104011,0.226301,0.260605,0.260605,0.222489,0.175983,0.247152,0.205152,...,1.000137e+00,1.000137e+00,1.000137e+00,1.000137e+00,1.000137e+00,1.000137e+00,1.000137e+00,0.156346,0.169589,0.212924
min,0.000000,1.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,-2.352020e+00,-3.720788e+00,-2.943049e+00,-1.707577e+00,-3.323522e+00,-1.889062e+00,-1.985818e+00,0.000000,0.000000,0.000000
25%,1.000000,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,-6.863816e-01,-7.258178e-01,-7.928291e-01,-8.268712e-01,-7.886676e-01,-8.262286e-01,-4.342076e-01,0.000000,0.125000,0.000000
50%,4.000000,3.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,-2.012609e-02,1.726734e-01,-3.150024e-01,5.383427e-02,-2.253666e-01,-1.176729e-01,-4.342076e-01,0.125000,0.250000,0.000000
75%,9.000000,4.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,...,6.461294e-01,4.721705e-01,6.406508e-01,5.822576e-01,6.195849e-01,7.680217e-01,8.299572e-02,0.250000,0.375000,0.333333
max,30.000000,6.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,1.000000,...,2.478332e+00,4.964626e+00,3.507611e+00,3.400515e+00,4.844343e+00,3.247967e+00,1.870231e+01,1.000000,1.000000,1.000000



--- First 5 Records ---


,CR,Size_Rank,is_swarm,Type_aberration,Type_beast,Type_construct,Type_dragon,Type_elemental,Type_fiend,Type_giant,...,Strength_Z,Dexterity_Z,Constitution_Z,Intelligence_Z,Wisdom_Z,Charisma_Z,Max_Speed_Z,Trait_Count_MM,Action_Count_MM,Movement_Count_MM
0,8.000,4,0,0,0,0,0,0,0,1,...,0.979257,-1.624309,0.401738,-0.826871,-0.788668,-0.649090,0.082996,0.125,1.000,0.000000
1,0.250,3,0,0,0,0,0,0,0,0,...,-0.852945,0.172673,-1.270656,0.053834,-0.225367,-0.117673,0.600199,0.000,0.250,0.333333
2,4.000,3,0,0,0,0,0,1,0,0,...,-0.852945,0.771668,-0.792829,0.406116,1.182886,0.059466,0.600199,0.000,0.250,0.333333
3,0.125,3,0,0,0,0,0,0,0,0,...,-0.852945,0.172673,-1.270656,0.053834,-0.225367,-0.117673,0.600199,0.125,0.250,0.333333
4,0.250,3,0,0,0,0,0,1,0,0,...,-0.852945,0.172673,-0.792829,0.053834,-0.225367,-0.117673,0.600199,0.000,0.125,0.333333


# Data Splitting Results
We utilize two splitting strategies depending on the intended model architecture. All splits are stratified by CR to maintain balance.

### Strategy 1: Train-Test Split (80/20)
*Used for initial benchmarking and standard regression models.*
* **Training Set (80%):** 2,495 monsters (The "Knowledge Base")
* **Test Set (20%):** 624 monsters (The "Final Exam")

In [2]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Remove duplicates based on unique records audit
df_unique = preprocessed_df.drop_duplicates()

# Define Features (X) and Target (y)
X = df_unique.drop(columns=['CR'])
y = df_unique['CR']

# Create bins for stratification to ensure representation of rare CR values
bins = [0, 0.5, 2, 5, 10, 20, 100]
y_binned = pd.cut(y, bins=bins, labels=False, include_lowest=True)

# Perform Train-Test Split (80/20)
X_train_80, X_test_20, y_train_80, y_test_20 = train_test_split(
    X, y, 
    test_size=0.20, 
    random_state=42, 
    stratify=y_binned
)

# Print verification
print(f"Total Unique Samples: {len(df_unique)}")
print(f"Training Set (80%): {len(X_train_80)} monsters")
print(f"Test Set (20%):     {len(X_test_20)} monsters")

Total Unique Samples: 3119
Training Set (80%): 2495 monsters
Test Set (20%):     624 monsters


### Strategy 2: Train-Validation-Test Split (70/15/15)
*Used for Neural Networks to allow for real-time hyperparameter tuning.*
* **Training Set (70%):** 2,183 monsters
* **Validation Set (15%):** 468 monsters (To prevent overfitting)
* **Test Set (15%):** 468 monsters (Final unbiased evaluation)

In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split

# Load and prepare data
df = pd.read_csv('../../data/proposed/preprocessed_bestiary.csv')
df_unique = df.drop_duplicates()

X = df_unique.drop(columns=['CR'])
y = df_unique['CR']

# Define bins for stratification
bins = [0, 0.5, 2, 5, 10, 20, 100]
y_binned = pd.cut(y, bins=bins, labels=False, include_lowest=True)

# Split 15% for final Test Set
X_temp, X_test_15, y_temp, y_test_15 = train_test_split(
    X, y, 
    test_size=0.15, 
    random_state=42, 
    stratify=y_binned
)

# Split remaining data into Train (70%) and Validation (15%)
y_temp_binned = pd.cut(y_temp, bins=bins, labels=False, include_lowest=True)

X_train_70, X_val_15, y_train_70, y_val_15 = train_test_split(
    X_temp, y_temp, 
    test_size=0.1765, # 0.15 / 0.85
    random_state=42, 
    stratify=y_temp_binned
)

# Print verification
print(f"Training Set (70%):   {len(X_train_70)} monsters")
print(f"Validation Set (15%): {len(X_val_15)} monsters")
print(f"Test Set (15%):       {len(X_test_15)} monsters")

Training Set (70%):   2183 monsters
Validation Set (15%): 468 monsters
Test Set (15%):       468 monsters


# Final Validation
Before moving to Exploratory Data Analysis (EDA), we perform a final audit against our initial IBM quality pillars:

* **Accuracy & Consistency Check:** All numerical columns have been verified as valid floats or integers.
* **Completeness Check:** A final sweep confirms zero remaining null values in our feature set.
* **Uniqueness Check:** The dataset contains 3,119 unique entries, ensuring a diverse and de-duplicated population for machine learning.

**Conclusion:** With these checks complete, the data is now considered "high quality" and ready for deeper exploration.

## About the New Dataset
This dataset contains **3,119 unique monsters** with **28 engineered features**. It serves as a distilled "Combat DNA" repository, gathering the stats used to build creatures for D&D 5e. By training a model on these features, we can predict a monster's Challenge Rating (CR) to help DMs validate homebrew content quickly.

**Dataset Source:** [DnD 5e Bestiary](https://5e.tools/bestiary.html)

---

### Final Feature Bestiary

Here are the **28 engineered features** that represent a monster's "Combat DNA." These features have been processed into numerical formats to allow the machine learning model to identify patterns in creature power levels.

#### Target Variable
1. **CR (Challenge Rating)**: A metric used to gauge the level of difficulty or danger a creature poses. The numerical value representing the creature's overall power level. This is the **dependent variable** our model intends to predict. Example: `30.0 (The Tarrasque)`

#### Health & Defense
2. **Log_Avg_HP**: Compressed survivability signal. Hit Points transformed using $\ln(HP + 1)$ to tame exponential health growth at high levels, making the relationship more linear for the model. Example: `4.8 (representing ~120 HP)`
3. **AC_Z**: Relative armor effectiveness. The standardized Z-score of Armor Class, centered around a mean of ~14. It allows the model to see how "hard to hit" a monster is compared to the average. Example: `1.2 (High AC)`

#### Standardized Ability Scores
4. **Strength_Z**: Physical power and force. The Z-score representing a creature's capacity for physical tasks and melee damage relative to the dataset average. Example: `2.5 (Giant-level strength)`
5. **Dexterity_Z**: Agility, reflexes, and coordination. The Z-score representing speed, balance, and accuracy relative to the average. Example: `1.5`
6. **Constitution_Z**: Physical health and stamina. The Z-score representing endurance, which directly dictates total Hit Points. Example: `0.8`
7. **Intelligence_Z**: Logic, reasoning, and memory. The Z-score representing mental acuity and the ability to perform complex calculations. Example: `-0.5`
8. **Wisdom_Z**: Intuition, insight, and awareness. The Z-score representing situational awareness and resistance to mental influence. Example: `0.0`
9. **Charisma_Z**: Force of personality and social influence. The Z-score representing a creature's magnetism and internal magical power. Example: `2.0`

#### Taxonomy & Role
10. **Size_Rank**: Physical scale of the creature. An ordinal rank from 1 (Tiny) to 6 (Gargantuan) that dictates the area a creature occupies in combat. Example: `3 (Medium)`
11. **is_swarm**: Tactical state flag. A binary indicator (1 or 0) for creatures acting as a single unit composed of multiple smaller creatures. Example: `1 (Swarm of Bats)`
12. **Is_Legendary**: Boss-tier status. A binary flag (1 or 0) based on the presence of legendary actions, marking the creature as a major threat. Example: `1 (Ancient Dragon)`
13. **Is_Spellcaster**: Magical capacity. A binary flag (1 or 0) for monsters with spellcasting traits or accessible spell slots. Example: `1 (Lich)`

#### Creature Types
14. **Type_aberration**: Binary flag (1/0) indicating an alien, mind-warping entity from the Far Realm.
15. **Type_beast**: Binary flag (1/0) indicating a natural creature of the fantasy world (e.g., bears, wolves).
16. **Type_construct**: Binary flag (1/0) indicating an artificial being created through magic or technology.
17. **Type_dragon**: Binary flag (1/0) indicating a powerful, reptilian, and often winged magical creature.
18. **Type_elemental**: Binary flag (1/0) indicating a being composed of fundamental forces like fire, air, or earth.
19. **Type_fiend**: Binary flag (1/0) indicating a creature from the lower planes, such as demons or devils.
20. **Type_giant**: Binary flag (1/0) indicating a humanoid creature of immense size and strength.
21. **Type_humanoid**: Binary flag (1/0) indicating the primary peoples of the world (e.g., elves, dwarves, humans).
22. **Type_monstrosity**: Binary flag (1/0) indicating unnatural creatures, often the result of experiments or curses.
23. **Type_undead**: Binary flag (1/0) indicating creatures that were once alive but now exist through dark magic.
24. **Type_other**: Binary flag (1/0) for less common types like Celestials, Plants, Oozes, or Fey.

#### Mobility & Complexity
25. **Max_Speed_Z**: Relative mobility. The standardized Z-score of the highest movement mode (Walk, Fly, Swim, etc.), weighing how fast a creature is compared to others.
26. **Movement_Count_MM**: Mobility diversity. Min-Max scaled count (0 to 1) of different movement modes. High values indicate versatile monsters that can fly and swim.
27. **Trait_Count_MM**: Passive complexity. Min-Max scaled count (0 to 1) of special passive traits, such as "Magic Resistance" or "Spider Climb."
28. **Action_Count_MM**: Offensive versatility. Min-Max scaled count (0 to 1) of available combat actions, representing how many options a creature has on its turn.

In [ ]:
# Save 80/20 splits (for regression models)
X_train_80.to_csv('../../data/proposed/final_split/X_train.csv', index=False)
X_test_20.to_csv('../../data/proposed/final_split/X_test.csv', index=False)
y_train_80.to_csv('../../data/proposed/final_split/y_train.csv', index=False)
y_test_20.to_csv('../../data/proposed/final_split/y_test.csv', index=False)

# Save 70/15/15 splits (for neural network)
X_train_70.to_csv('../../data/proposed/final_split/X_train_nn.csv', index=False)
X_val_15.to_csv('../../data/proposed/final_split/X_val_nn.csv', index=False)
X_test_15.to_csv('../../data/proposed/final_split/X_test_nn.csv', index=False)
y_train_70.to_csv('../../data/proposed/final_split/y_train_nn.csv', index=False)
y_val_15.to_csv('../../data/proposed/final_split/y_val_nn.csv', index=False)
y_test_15.to_csv('../../data/proposed/final_split/y_test_nn.csv', index=False)